File: src/sim/visibilityGraph.ts
import { expandPolygon, pointInPolygon, pointSegmentDistance, segmentsCross, distance, type Point } from './geometry';

/** Tolerance for "this point sits on the hull outline rather than inside it". */
const BOUNDARY_EPSILON = 1e-6;
import type { CsrGraph } from './dijkstra';
import { convexDecompose } from './convexDecompose';
import type { Wall } from '../state/model';

/**
 * How much further out the graph nodes sit than the blocking boundary.
 *
 * Without it a node lands exactly *on* the hull it belongs to, where
 * point-in-polygon is a coin flip and half the surrounding lattice cells are
 * illegal. Pedestrians could then never legally stand on their own waypoint and
 * would stall against the corner, jittering. A couple of units of clearance makes
 * every node a position a pedestrian can actually occupy.
 */
export const NODE_MARGIN = 2;

export interface Obstacle {
  wallId: number;
  /** Index of this convex part among all obstacles, used for ring adjacency. */
  partId: number;
  /** One convex part of the wall, pushed out by the pedestrian radius. */
  hull: Point[];
  bbox: { minX: number; minY: number; maxX: number; maxY: number };
}

/**
 * A whole wall's convex hull, expanded by the radius: the shape the original drew
 * dashed. Kept as a broad phase -- anything that misses the shell cannot touch any
 * of the wall's convex parts, so the per-part work is skipped entirely.
 */
export interface WallShell {
  wallId: number;
  hull: Point[];
  bbox: { minX: number; minY: number; maxX: number; maxY: number };
}

/** A wall's shell together with its convex parts, so the broad phase needs no lookup. */
export interface WallGroup {
  shell: WallShell | null;
  parts: Obstacle[];
}

export interface Blockers {
  obstacles: Obstacle[];
  shells: WallShell[];
  groups: WallGroup[];
}

export interface VisibilityGraph extends Blockers {
  nodes: Point[];
  /** Which wall each node's part belongs to; goals are seeded by this. */
  nodeWall: Int32Array;
  /** Which convex part each node came from. */
  nodePart: Int32Array;
  /** Position of each node within its own hull ring, for the adjacency rule. */
  nodeRingIndex: Int32Array;
  ringLength: Int32Array;
  csr: CsrGraph;
}

function bboxOf(poly: Point[]) {
  let minX = Infinity, minY = Infinity, maxX = -Infinity, maxY = -Infinity;
  for (const [x, y] of poly) {
    if (x < minX) minX = x;
    if (x > maxX) maxX = x;
    if (y < minY) minY = y;
    if (y > maxY) maxY = y;
  }
  return { minX, minY, maxX, maxY };
}

function segmentMissesBox(a: Point, b: Point, box: Obstacle['bbox']): boolean {
  return (Math.max(a[0], b[0]) < box.minX) || (Math.min(a[0], b[0]) > box.maxX)
    || (Math.max(a[1], b[1]) < box.minY) || (Math.min(a[1], b[1]) > box.maxY);
}

/**
 * Whether a straight walk from `a` to `b` is possible.
 *
 * The hulls are already expanded by the pedestrian radius, so clearance is built
 * into the geometry -- which is why this needs no distance-to-wall check, unlike
 * Map.isVisible(), which tested raw wall lines and then re-checked every corner
 * against the pedestrian radius.
 *
 * Two rules:
 *  1. The segment must not properly cross any hull edge.
 *  2. Its midpoint must not fall strictly inside a hull. This is what rejects a
 *     chord between two non-adjacent corners of the *same* convex hull, which
 *     crosses no edge yet passes straight through the building.
 *
 * The midpoint is exempt when it lies *on* the outline. Walking from one hull
 * corner to the next puts the midpoint exactly on that edge, and ray casting on
 * the boundary is a coin flip -- without the exemption, agents cannot walk along
 * a wall they are standing against and park on the corner forever.
 */
export function isVisible(a: Point, b: Point, blockers: Blockers): boolean {
  const mid: Point = [(a[0] + b[0]) / 2, (a[1] + b[1]) / 2];

  // Broad phase: a wall whose shell the segment never touches cannot be hit by
  // any of its convex parts, so skip the whole group. Walls are pre-grouped with
  // their parts so this allocates nothing -- it runs for every candidate edge and
  // every legality check, so a per-call Set here costs more than it saves.
  for (const group of blockers.groups) {
    const shell = group.shell;
    if (shell) {
      if (segmentMissesBox(a, b, shell.bbox)) continue;
      if (!touchesHull(a, b, mid, shell.hull)) continue;
    }
    for (const ob of group.parts) {
      if (segmentMissesBox(a, b, ob.bbox)) continue;
      const hull = ob.hull;
      let onOutline = false;
      for (let i = 0, n = hull.length; i < n; i++) {
        const p = hull[i];
        const q = hull[(i + 1) % n];
        if (segmentsCross(a, b, p, q)) return false;
        if (!onOutline && pointSegmentDistance(p, q, mid) <= BOUNDARY_EPSILON) onOutline = true;
      }
      if (!onOutline && pointInPolygon(hull, mid)) return false;
    }
  }
  return true;
}

/** Whether a segment enters or crosses a convex outline at all. */
function touchesHull(a: Point, b: Point, mid: Point, hull: Point[]): boolean {
  for (let i = 0, n = hull.length; i < n; i++) {
    if (segmentsCross(a, b, hull[i], hull[(i + 1) % n])) return true;
  }
  return pointInPolygon(hull, a) || pointInPolygon(hull, b) || pointInPolygon(hull, mid);
}

/**
 * Builds the graph Dijkstra runs over: nodes are the corners of every wall's
 * convex hull expanded by the pedestrian radius, edges join corners that can see
 * each other.
 *
 * Construction is the naive O(n^2) pair sweep with a bounding-box reject per
 * obstacle, which is ample for the hundreds of corners a hand-drawn map produces.
 * The upgrade path if that ever stalls is Lee's rotational plane sweep; nothing
 * outside this function would need to change.
 */
export function buildVisibilityGraph(walls: Wall[], radius: number): VisibilityGraph {
  const obstacles: Obstacle[] = [];
  const nodes: Point[] = [];
  const nodeWall: number[] = [];
  const nodePart: number[] = [];
  const nodeRingIndex: number[] = [];
  const ringLength: number[] = [];

  // Each wall contributes one obstacle per convex part. For a convex wall that is
  // a single part identical to its hull; for a U or an L it is several, which is
  // what keeps the cavity walkable instead of hulled over.
  const shells: WallShell[] = [];
  // Node positions live on a slightly larger ring than the blocking hull, so that
  // every node is somewhere a pedestrian can legally stand.
  const nodeRings: Point[][] = [];

  for (const wall of walls) {
    if (wall.hull.length >= 3) {
      const shellHull = expandPolygon(wall.hull, radius);
      shells.push({ wallId: wall.id, hull: shellHull, bbox: bboxOf(shellHull) });
    }
    for (const poly of wall.polygons) {
      for (const part of convexDecompose(poly)) {
        if (part.length < 3) continue;
        const hull = expandPolygon(part, radius);
        obstacles.push({ wallId: wall.id, partId: obstacles.length, hull, bbox: bboxOf(hull) });
        nodeRings.push(expandPolygon(part, radius + NODE_MARGIN));
      }
    }
  }
  const groups: WallGroup[] = walls.map((w) => ({
    shell: shells.find((sh) => sh.wallId === w.id) ?? null,
    parts: obstacles.filter((ob) => ob.wallId === w.id),
  }));
  const blockers: Blockers = { obstacles, shells, groups };

  // A corner swallowed by another building's hull is not standable, so drop it.
  for (let o = 0; o < obstacles.length; o++) {
    const ob = obstacles[o];
    const ring = nodeRings[o];
    for (let i = 0; i < ring.length; i++) {
      const p = ring[i];
      let blocked = false;
      for (let k = 0; k < obstacles.length && !blocked; k++) {
        if (k === o) continue;
        if (pointInPolygon(obstacles[k].hull, p)) blocked = true;
      }
      if (blocked) continue;
      nodes.push(p);
      nodeWall.push(ob.wallId);
      nodePart.push(ob.partId);
      nodeRingIndex.push(i);
      ringLength.push(ring.length);
    }
  }

  // Pairwise visibility.
  const n = nodes.length;
  const neighbours: number[][] = Array.from({ length: n }, () => []);
  const costs: number[][] = Array.from({ length: n }, () => []);

  for (let i = 0; i < n; i++) {
    for (let j = i + 1; j < n; j++) {
      // Every pair gets a real visibility test, including two corners of the same
      // convex part.
      //
      // There used to be a shortcut here: corners of one part were taken as
      // mutually visible if they were ring neighbours, on the reasoning that
      // walking around your own convex hull is always safe. It is not. A wall
      // made of several parts can have one part lying across another's edge --
      // a vertical bar crossing the underside of a horizontal one -- and the
      // shortcut let the graph run an edge straight through it, so pedestrians
      // walked into a wall and jammed against it.
      //
      // The shortcut existed to dodge the boundary ambiguity of a segment running
      // exactly along the hull it belongs to. That is no longer a concern: nodes
      // sit on a ring NODE_MARGIN outside the blocking hull, so the segment
      // between two of them is strictly outside and decides cleanly.
      if (!isVisible(nodes[i], nodes[j], blockers)) continue;
      const w = distance(nodes[i], nodes[j]);
      neighbours[i].push(j); costs[i].push(w);
      neighbours[j].push(i); costs[j].push(w);
    }
  }

  const offsets = new Int32Array(n + 1);
  for (let i = 0; i < n; i++) offsets[i + 1] = offsets[i] + neighbours[i].length;
  const targets = new Int32Array(offsets[n]);
  const weights = new Float32Array(offsets[n]);
  for (let i = 0, k = 0; i < n; i++) {
    for (let e = 0; e < neighbours[i].length; e++, k++) {
      targets[k] = neighbours[i][e];
      weights[k] = costs[i][e];
    }
  }

  return {
    nodes,
    nodeWall: Int32Array.from(nodeWall),
    nodePart: Int32Array.from(nodePart),
    nodeRingIndex: Int32Array.from(nodeRingIndex),
    ringLength: Int32Array.from(ringLength),
    obstacles,
    shells,
    groups,
    csr: { nodeCount: n, offsets, targets, weights },
  };
}

/** Node indices belonging to a given wall -- the seed set for a goal's Dijkstra run. */
export function nodesOfWall(graph: VisibilityGraph, wallId: number): number[] {
  const out: number[] = [];
  for (let i = 0; i < graph.nodeWall.length; i++) {
    if (graph.nodeWall[i] === wallId) out.push(i);
  }
  return out;
}

File: ios/Sources/WalkySim/VisibilityGraph.swift
import Foundation

/// Ports `src/sim/visibilityGraph.ts`.
///
/// Nodes are the corners of every wall's convex parts expanded by the
/// pedestrian radius; edges join corners that can see each other.

/// Tolerance for "this point sits on the hull outline rather than inside it".
private let BOUNDARY_EPSILON = 1e-6

/// How much further out the graph nodes sit than the blocking boundary.
///
/// Without it a node lands exactly *on* the hull it belongs to, where
/// point-in-polygon is a coin flip and half the surrounding lattice cells are
/// illegal -- pedestrians could never legally stand on their own waypoint and
/// would stall against the corner, jittering.
public let NODE_MARGIN: Double = 2

/// How much wider than the pedestrian radius a wall's shell is expanded.
///
/// The shell is only a reject, so it is only sound while it contains every
/// part. A needle part inside the hull can have a far sharper corner than the
/// hull does, so its own cut corner may reach further out than the hull's.
///
/// Baked as a literal rather than computed: in JS this is a module-init
/// `Math.hypot(MITER_LIMIT, 1)`, and while `jsHypot` would give the same bits,
/// a constant that must match across two languages is better written down once
/// than derived twice.
private let SHELL_SLACK: Double = 2.23606797749978969   // hypot(2, 1) = sqrt(5)

public struct BBox {
  public var minX: Double, minY: Double, maxX: Double, maxY: Double
}

public struct Obstacle {
  public var wallId: Int
  /// Index of this convex part among all obstacles, used for ring adjacency.
  public var partId: Int
  /// One convex part of the wall, pushed out by the pedestrian radius.
  public var hull: [Point]
  public var bbox: BBox
}

/// A whole wall's convex hull, expanded: the broad phase for that wall.
public struct WallShell {
  public var wallId: Int
  public var hull: [Point]
  public var bbox: BBox
}

/// A wall's shell together with its parts, so the broad phase needs no lookup.
public struct WallPartGroup {
  public var shell: WallShell?
  public var parts: [Obstacle]
}

public struct Blockers {
  public var obstacles: [Obstacle]
  public var shells: [WallShell]
  public var groups: [WallPartGroup]
}

public struct VisibilityGraph {
  public var nodes: [Point]
  /// Which wall each node's part belongs to; goals are seeded by this.
  public var nodeWall: [Int32]
  /// Which convex part each node came from.
  public var nodePart: [Int32]
  /// Position of each node within its own hull ring, for the adjacency rule.
  public var nodeRingIndex: [Int32]
  public var ringLength: [Int32]
  public var csr: CsrGraph
  public var blockers: Blockers

  public var obstacles: [Obstacle] { blockers.obstacles }
  public var shells: [WallShell] { blockers.shells }
}

private func bboxOf(_ poly: [Point]) -> BBox {
  var minX = Double.infinity, minY = Double.infinity
  var maxX = -Double.infinity, maxY = -Double.infinity
  for p in poly {
    if p.x < minX { minX = p.x }
    if p.x > maxX { maxX = p.x }
    if p.y < minY { minY = p.y }
    if p.y > maxY { maxY = p.y }
  }
  return BBox(minX: minX, minY: minY, maxX: maxX, maxY: maxY)
}

@inline(__always)
private func segmentMissesBox(_ a: Point, _ b: Point, _ box: BBox) -> Bool {
  (jsMax(a.x, b.x) < box.minX) || (jsMin(a.x, b.x) > box.maxX)
    || (jsMax(a.y, b.y) < box.minY) || (jsMin(a.y, b.y) > box.maxY)
}

/// Whether a straight walk from `a` to `b` is possible.
///
/// The hulls are already expanded by the pedestrian radius, so clearance is
/// built into the geometry. Two rules: the segment must not properly cross any
/// hull edge, and its midpoint must not fall strictly inside a hull -- which is
/// what rejects a chord between two non-adjacent corners of the *same* hull,
/// crossing no edge yet passing straight through the building.
///
/// The midpoint is exempt when it lies *on* the outline: walking from one hull
/// corner to the next puts the midpoint exactly on that edge, and ray casting
/// on the boundary is a coin flip.
public func isVisible(_ a: Point, _ b: Point, _ blockers: Blockers) -> Bool {
  let mid = Point((a.x + b.x) / 2, (a.y + b.y) / 2)

  for group in blockers.groups {
    if let shell = group.shell {
      if segmentMissesBox(a, b, shell.bbox) { continue }
      if !touchesHull(a, b, mid, shell.hull) { continue }
    }
    for ob in group.parts {
      if segmentMissesBox(a, b, ob.bbox) { continue }
      let hull = ob.hull
      var onOutline = false
      let n = hull.count
      for i in 0..<n {
        let p = hull[i]
        let q = hull[(i + 1) % n]
        if segmentsCross(a, b, p, q) { return false }
        if !onOutline && pointSegmentDistance(p, q, mid) <= BOUNDARY_EPSILON { onOutline = true }
      }
      if !onOutline && pointInPolygon(hull, mid) { return false }
    }
  }
  return true
}

/// Whether a segment enters or crosses a convex outline at all.
private func touchesHull(_ a: Point, _ b: Point, _ mid: Point, _ hull: [Point]) -> Bool {
  let n = hull.count
  for i in 0..<n {
    if segmentsCross(a, b, hull[i], hull[(i + 1) % n]) { return true }
  }
  return pointInPolygon(hull, a) || pointInPolygon(hull, b) || pointInPolygon(hull, mid)
}

/// Builds the graph Dijkstra runs over.
///
/// The naive O(n^2) pair sweep with a bounding-box reject per obstacle, which is
/// ample for the hundreds of corners a hand-drawn map produces.
public func buildVisibilityGraph(_ walls: [Wall], _ radius: Double) -> VisibilityGraph {
  var obstacles: [Obstacle] = []
  var nodes: [Point] = []
  var nodeWall: [Int32] = []
  var nodePart: [Int32] = []
  var nodeRingIndex: [Int32] = []
  var ringLength: [Int32] = []

  var shells: [WallShell] = []
  // Node positions live on a slightly larger ring than the blocking hull, so
  // that every node is somewhere a pedestrian can legally stand.
  var nodeRings: [[Point]] = []

  for wall in walls {
    // A wall of fewer than three distinct points has no hull to reject against;
    // that costs the broad phase, never an obstacle.
    if wall.hull.count >= 3 {
      let shellHull = expandPolygon(wall.hull, radius * SHELL_SLACK)
      shells.append(WallShell(wallId: wall.id, hull: shellHull, bbox: bboxOf(shellHull)))
    }
    for poly in wall.polygons {
      for part in convexDecompose(poly) {
        if part.count < 3 { continue }
        let hull = expandPolygon(part, radius)
        obstacles.append(Obstacle(wallId: wall.id, partId: obstacles.count,
                                  hull: hull, bbox: bboxOf(hull)))
        nodeRings.append(expandPolygon(part, radius + NODE_MARGIN))
      }
    }
  }

  let groups: [WallPartGroup] = walls.map { w in
    WallPartGroup(shell: shells.first { $0.wallId == w.id },
                  parts: obstacles.filter { $0.wallId == w.id })
  }
  let blockers = Blockers(obstacles: obstacles, shells: shells, groups: groups)

  // A corner swallowed by another building's hull is not standable, so drop it.
  for o in 0..<obstacles.count {
    let ob = obstacles[o]
    let ring = nodeRings[o]
    for i in 0..<ring.count {
      let p = ring[i]
      var blocked = false
      var k = 0
      while k < obstacles.count && !blocked {
        if k == o { k += 1; continue }
        if pointInPolygon(obstacles[k].hull, p) { blocked = true }
        k += 1
      }
      if blocked { continue }
      nodes.append(p)
      nodeWall.append(Int32(ob.wallId))
      nodePart.append(Int32(ob.partId))
      nodeRingIndex.append(Int32(i))
      ringLength.append(Int32(ring.count))
    }
  }

  // Pairwise visibility. Every pair gets a real test, including two corners of
  // the same convex part: a wall made of several parts can have one lying
  // across another's edge, and the old ring-neighbour shortcut ran an edge
  // straight through it.
  let n = nodes.count
  var neighbours = [[Int32]](repeating: [], count: n)
  var costs = [[Double]](repeating: [], count: n)

  for i in 0..<n {
    for j in (i + 1)..<n {
      if !isVisible(nodes[i], nodes[j], blockers) { continue }
      let w = distance(nodes[i], nodes[j])
      neighbours[i].append(Int32(j)); costs[i].append(w)
      neighbours[j].append(Int32(i)); costs[j].append(w)
    }
  }

  var offsets = [Int32](repeating: 0, count: n + 1)
  for i in 0..<n { offsets[i + 1] = offsets[i] + Int32(neighbours[i].count) }
  let edgeCount = Int(offsets[n])
  var targets = [Int32](repeating: 0, count: edgeCount)
  var weights = [Float](repeating: 0, count: edgeCount)
  var k = 0
  for i in 0..<n {
    for e in 0..<neighbours[i].count {
      targets[k] = neighbours[i][e]
      weights[k] = Float(costs[i][e])
      k += 1
    }
  }

  return VisibilityGraph(
    nodes: nodes, nodeWall: nodeWall, nodePart: nodePart,
    nodeRingIndex: nodeRingIndex, ringLength: ringLength,
    csr: CsrGraph(nodeCount: n, offsets: offsets, targets: targets, weights: weights),
    blockers: blockers)
}

/// Node indices belonging to a wall -- the seed set for a goal's Dijkstra run.
public func nodesOfWall(_ graph: VisibilityGraph, _ wallId: Int) -> [Int] {
  var out: [Int] = []
  for i in 0..<graph.nodeWall.count where graph.nodeWall[i] == Int32(wallId) { out.append(i) }
  return out
}
